### ISSUE TITLE
Support auto-discovery

### ISSUE THREAD

**Issue by danielperna84** [SukramJ/hahomematic#224]
**Is your feature request related to a problem? Please describe.**
HomeMatic has to be set up manually by going to the integrations menu. It would be nice if one day the user simply would get a hint that the integration is available for setup via auto-discovery.

**Describe the solution you'd like**
The CCU (Debmatic in my case) has an endpoint like `http://1.2.3.4/upnp/basic_dev.cgi`, and I remember RaspberryMatic having something similar.

The response looks like this:

```xml
<root xmlns="urn:schemas-upnp-org:device-1-0">
<specVersion>
<major>1</major>
<minor>0</minor>
</specVersion>
<URLBase>http://:30080/upnp/</URLBase>
<device>
<deviceType>urn:schemas-upnp-org:device:Basic:1</deviceType>
<presentationURL/>
<friendlyName>HomeMatic Central - myhostname</friendlyName>
<manufacturer>EQ3</manufacturer>
<manufacturerURL>http://www.homematic.com</manufacturerURL>
<modelDescription>HomeMatic Central 3014F711A0001F58A9A7264E</modelDescription>
<modelName>HomeMatic Central</modelName>
<UDN>uuid:upnp-BasicDevice-1_0-3014F711A0001F58A9A7264E</UDN>
<UPC>123456789002</UPC>
<serviceList>
<service>
<serviceType>urn:schemas-upnp-org:service:dummy:1</serviceType>
<serviceId>urn:upnp-org:serviceId:dummy1</serviceId>
<controlURL/>
<eventSubURL/>
<SCPDURL/>
</service>
</serviceList>
</device>
</root>
```

I also have seen posts in the HM forum about the sspd-service being available on the CCU. Hence I believe it should be possible to perform auto-discovery.

**Describe alternatives you've considered**
Setting it up manually. 😅

**Additional context**
This is a low priority issue just to remind ourselves (and others watching the repo), that this feature could be integrated some day. I don't know if it's possible to already add auto-discovery support while we are still testing this as a custom component, or if it has to be a _real_ one. In any case,  we can use this issue to keep track of the feature IF we decide to implement this.

**Comment by SukramJ** [comment]
Adding this to the custom_component is really simple, but what are the relevant data that we could use in this response?

**Comment by danielperna84** [comment]
I think it would be useful to see the responses of other CCUs so we can compare this. At first glance I would say that these fields are interesting:

```xml
<manufacturer>EQ3</manufacturer>
<manufacturerURL>http://www.homematic.com</manufacturerURL>
<modelDescription>HomeMatic Central 3014F711A0001F58A9A7264E</modelDescription>
<modelName>HomeMatic Central</modelName>
```

That being said, I don't know if the way HA does auto-discovery actually works be looking at this. But a assume it's safe to assume a device responding with _EQ3_ and multiple _HomeMatic_ strings is a CCU.

Using `gssdp-discover` I get the following for my machine:

```
resource available
  USN:      uuid:upnp-BasicDevice-1_0-3014F711A0001F58A9A7264E::upnp:rootdevice
  Location: http://192.168.1.20/upnp/basic_dev.cgi
```

I agree that this alone might not be sufficient. We would have to check if this `3014F...` ID is unique per CCU, or if this actually is usable to identify all CCUs (of a certain type / firmware).

__Edit:__  
Well, according to [this](https://github.com/jens-maus/RaspberryMatic/commit/d75c599501561ae11d7325a21318801fe81dda59) commit that ID probably really is the serial number of the device. So we can't identify by this I guess. That would leave us with the `uuid:upnp-BasicDevice-1_0-*::upnp:rootdevice` if all we get in the first step is that information. If this can be used to verify it's a CCU by looking into the XML in a second step, I guess that would be an option.

__Edit 2:__  
Well, at least in my network no other device has a location with `/upnp/basic_dev.cgi`. So this could hint at a CCU as well.

**Comment by SukramJ** [comment]
3014F... is the SN of the RF-module

**Comment by SukramJ** [comment]
This is what HA discovers:

```
SsdpServiceInfo(upnp={
  'deviceType': 'urn:schemas-upnp-org:device:Basic:1',
  'presentationURL': None,
  'friendlyName': 'HomeMatic Central - otto.XXXX.de',
  'manufacturer': 'EQ3',
  'manufacturerURL': 'http://www.homematic.com',
  'modelDescription': 'HomeMatic Central 3014F711AXXXXXXXXXXXXXX',
  'modelName': 'HomeMatic Central',
  'UDN': 'uuid:upnp-BasicDevice-1_0-3014F711AXXXXXXXXXXXXXX',
  'UPC': '123456789002',
  'serviceList': {
    'service': {
      'serviceType': 'urn:schemas-upnp-org:service:dummy:1',
      'serviceId': 'urn:upnp-org:serviceId:dummy1',
      'controlURL': None,
      'eventSubURL': None,
      'SCPDURL': None
    }
  }
}, 
ssdp_usn='uuid:upnp-BasicDevice-1_0-3014F711AXXXXXXXXXXXXXX::upnp:rootdevice', 
ssdp_st='upnp:rootdevice', 
ssdp_location='http://xx.xx.xx.xx/upnp/basic_dev.cgi', 
ssdp_nt=None, 
ssdp_udn='uuid:upnp-BasicDevice-1_0-3014F711AXXXXXXXXXXXXXX', 
ssdp_ext='', 
ssdp_server='HomeMatic', 
ssdp_headers={'CACHE-CONTROL': 'max-age=5000', 'EXT': '', 'SERVER': 'HomeMatic', 'ST': 'upnp:rootdevice', 
'USN': 'uuid:upnp-BasicDevice-1_0-3014F711AXXXXXXXXXXXXXX::upnp:rootdevice', '_location_original': 'http://xx.xx.xx.xx/upnp/basic_dev.cgi', 
'location': 'http://xx.xx.xx.xx/upnp/basic_dev.cgi', '_timestamp': datetime.datetime(2022, 1, 22, 18, 33, 49, 518821), 
'_host': 'xx.xx.xx.xx', '_port': 1900, '_udn': 'uuid:upnp-BasicDevice-1_0-3014F711AXXXXXXXXXXXXXX'}, 
x_homeassistant_matching_domains={
'hahm'
}, _warning_logged=False)
```

**Comment by danielperna84** [comment]
Well that looks promising. 👍

**Comment by SukramJ** [comment]
I created a branch upnp based on latest devel:

The config flow will detect an unconfigured/unignored instance.
`friendlyName` will be used for the **instance name** without the prefix `HomeMatic Central (- ) `
`HomeMatic Central - otto.XXXX.de` -> `otto.XXXX.de`
The `friendlyName` can be edited in CCU `Systemsteuerung/Netzwerkeinstellungen/Hostname`. Editing requires a restart of the CCU.

The **host** (IP) will be extracted from `ssdp_location`
`http://xx.xx.xx.xx/upnp/basic_dev.cgi` -> `xx.xx.xx.xx`

**instance name** and *host** (IP) will be used to prefill a config flow.

The required  change is to use the **ip_address** instead of the **instance_name** for the `unique_id` of the config_entry.
Users with an existing entry can **ignore** the newly detected instance or **reconfigure** the already configures instance. Reconfiguring will update the `unique_id` to the ip_address.

**Comment by danielperna84** [comment]
Is the IP used for the `unique_id` _technically_ relevant? I'm thinking about users who are using the hostname instead of the IP because of enabled TLS. And since IPs for a given hostname can change (f.ex. RaspberryMatic migrated from standalone to the HA-Addon, but hostname from DNS stays the same and only points to a different IP), I just want to make sure this won't cause issues.

**Comment by SukramJ** [comment]
> Is the IP used for the `unique_id` _technically_ relevant?

Yes, this is the only information we get from a manual config flow and a discovered one.
Whatever the user enters in the config_flow, i will convert the host to an ip_address that will be used as unique_id.
The entered host from the config_flow (FQDN or ip_address) will not be converted and be used for the communication to CCU.
If the user changes the CCUs ip_address. The CCU will be discovered again, but this can be fixed by clicking on reconfigure.

**Comment by danielperna84** [comment]
I have tried manually replacing `config_flow.py` and `manifest.json`, but I get the following error during HA startup:

```
Jan 23 16:11:10 hasstest2 hass[130]: 2022-01-23 16:11:10 ERROR (SyncWorker_3) [homeassistant.util.package] Unable to install package git+https://github.com/SukramJ/hahomematic.git@dev1#hahomematic: WARNING: Did not find branch or tag 'dev1', assuming revision or ref.
Jan 23 16:11:10 hasstest2 hass[130]: ERROR: Command errored out with exit status 1: git checkout -q dev1 Check the logs for full command output.
Jan 23 16:11:10 hasstest2 hass[130]: WARNING: You are using pip version 20.2.4; however, version 21.3.1 is available.
Jan 23 16:11:10 hasstest2 hass[130]: You should consider upgrading via the '/srv/homeassistant/bin/python3 -m pip install --upgrade pip' command.
Jan 23 16:11:11 hasstest2 hass[130]: 2022-01-23 16:11:11 ERROR (SyncWorker_5) [homeassistant.util.package] Unable to install package git+https://github.com/SukramJ/hahomematic.git@dev1#hahomematic: WARNING: Did not find branch or tag 'dev1', assuming revision or ref.
Jan 23 16:11:11 hasstest2 hass[130]: ERROR: Command errored out with exit status 1: git checkout -q dev1 Check the logs for full command output.
Jan 23 16:11:11 hasstest2 hass[130]: WARNING: You are using pip version 20.2.4; however, version 21.3.1 is available.
Jan 23 16:11:11 hasstest2 hass[130]: You should consider upgrading via the '/srv/homeassistant/bin/python3 -m pip install --upgrade pip' command.
Jan 23 16:11:12 hasstest2 hass[130]: 2022-01-23 16:11:12 ERROR (SyncWorker_0) [homeassistant.util.package] Unable to install package git+https://github.com/SukramJ/hahomematic.git@dev1#hahomematic: WARNING: Did not find branch or tag 'dev1', assuming revision or ref.
Jan 23 16:11:12 hasstest2 hass[130]: ERROR: Command errored out with exit status 1: git checkout -q dev1 Check the logs for full command output.
Jan 23 16:11:12 hasstest2 hass[130]: WARNING: You are using pip version 20.2.4; however, version 21.3.1 is available.
Jan 23 16:11:12 hasstest2 hass[130]: You should consider upgrading via the '/srv/homeassistant/bin/python3 -m pip install --upgrade pip' command.
Jan 23 16:11:12 hasstest2 hass[130]: 2022-01-23 16:11:12 ERROR (MainThread) [homeassistant] Error doing job: Task exception was never retrieved
Jan 23 16:11:12 hasstest2 hass[130]: Traceback (most recent call last):
Jan 23 16:11:12 hasstest2 hass[130]:   File "/srv/homeassistant/lib/python3.9/site-packages/homeassistant/data_entry_flow.py", line 203, in async_init
Jan 23 16:11:12 hasstest2 hass[130]:     flow, result = await task
Jan 23 16:11:12 hasstest2 hass[130]:   File "/srv/homeassistant/lib/python3.9/site-packages/homeassistant/data_entry_flow.py", line 221, in _async_init
Jan 23 16:11:12 hasstest2 hass[130]:     flow = await self.async_create_flow(handler, context=context, data=data)
Jan 23 16:11:12 hasstest2 hass[130]:   File "/srv/homeassistant/lib/python3.9/site-packages/homeassistant/config_entries.py", line 715, in async_create_flow
Jan 23 16:11:12 hasstest2 hass[130]:     await async_process_deps_reqs(self.hass, self._hass_config, integration)
Jan 23 16:11:12 hasstest2 hass[130]:   File "/srv/homeassistant/lib/python3.9/site-packages/homeassistant/setup.py", line 353, in async_process_deps_reqs
Jan 23 16:11:12 hasstest2 hass[130]:     await requirements.async_get_integration_with_requirements(
Jan 23 16:11:12 hasstest2 hass[130]:   File "/srv/homeassistant/lib/python3.9/site-packages/homeassistant/requirements.py", line 85, in async_get_integration_with_requirements
Jan 23 16:11:12 hasstest2 hass[130]:     await _async_process_integration(hass, integration, done)
Jan 23 16:11:12 hasstest2 hass[130]:   File "/srv/homeassistant/lib/python3.9/site-packages/homeassistant/requirements.py", line 101, in _async_process_integration
Jan 23 16:11:12 hasstest2 hass[130]:     await async_process_requirements(
Jan 23 16:11:12 hasstest2 hass[130]:   File "/srv/homeassistant/lib/python3.9/site-packages/homeassistant/requirements.py", line 164, in async_process_requirements
Jan 23 16:11:12 hasstest2 hass[130]:     await _async_process_requirements(
Jan 23 16:11:12 hasstest2 hass[130]:   File "/srv/homeassistant/lib/python3.9/site-packages/homeassistant/requirements.py", line 196, in _async_process_requirements
Jan 23 16:11:12 hasstest2 hass[130]:     raise RequirementsNotFound(name, [req])
Jan 23 16:11:12 hasstest2 hass[130]: homeassistant.requirements.RequirementsNotFound: Requirements for hahm not found: ['git+https://github.com/SukramJ/hahomematic.git@dev1#hahomematic'].
```

What's the proper way of testing this? HACS doesn't provide an option to select a different branch as far as I can tell. 🤔

**Comment by SukramJ** [comment]
Option 1 Fix manifest:
```
  "requirements": ["hahomematic==0.26.0"],
  "requirements1": ["git+https://github.com/SukramJ/hahomematic.git@dev1#hahomematic"],
```

Option 2 I merge it to devel.
Then you can select "erneut herunterladen" in hacs and select devel 

The users will not be affected.

**Comment by danielperna84** [comment]
I just wanted to test it before merging it. 😉 That being said, it worked flawlessly. 👍

**Comment by SukramJ** [comment]
Sounds good. Can be merged?

Releases are based on tags, so it is possible to create a pre-release based on a different branch.

**Comment by SukramJ** [comment]
Could you please update the docs regarding auto-discovery?

**Comment by danielperna84** [comment]
> Sounds good. Can be merged?

Yup.

> Could you please update the docs regarding auto-discovery?

Yup. I'll do that on the upnp branch, ok?

__Edit:__ Done, see: https://github.com/danielperna84/custom_homematic/commit/2dbe491673e14924accefc04ae1361a1d8161e49#diff-b335630551682c19a781afebcf4d07bf978fb1f8ac04c6bf87428ed5106870f5

__Edit 2:__ plus the commit after that for the typo.

**Comment by SukramJ** [comment]
Released with 0.26.2
